flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/TaskManagerLocation.java

```diff
@@ -150,28 +165,48 @@ public String addressString() {
 
 	/**
 	 * Returns the fully-qualified domain name the TaskManager. If the name could not be
-	 * determined, the return value will be a textual representation of the TaskManager's IP address.
+	 * determined or {@link #retrieveHostName} is set to false,
+	 * the return value will be a textual representation of the TaskManager's IP address.
 	 * 
 	 * @return The fully-qualified domain name of the TaskManager.
 	 */
 	public String getFQDNHostname() {
+		// Lazily retrieve FQDN host name of this TaskManager
+		if (fqdnHostName == null) {
+			if (!retrieveHostName) {
+				fqdnHostName = inetAddress.getHostAddress();
+			} else {
+				fqdnHostName = getFqdnHostName(inetAddress);
+			}
+		}
```

Comment:
I am wondering whether it wouldn't make sense to move the resolution logic out of the `TaskManagerLocation` class in order to make more like a simple data transfer object. For example, we could have a method `fromUnresolvedLocation(UnresolvedTaskManagerLocation, ResolutionMode)` which either resolves the hostname or not depending on `ResolutionMode`.

As a second step we could think about whether we want to support laziness. If this is the case, then we could provide a `Supplier<String> fqdnHostNameSupplier` which we call to give us the fqdn hostname and then store it some field.

Separating the `TaskManagerLocation` from the way it is resolved might simplify the individual classes a bit.

Reply:
Thanks Till for the valuable suggestions, and I agree that by extracting the hostname retrieval code into some Supplier-like classes could make the logics simpler and more intuitive.